File: backend/handlers/federation.go
package handlers

import (
	"context"
	"fmt"
	"time"

	"github.com/kutup/backend/services"
	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
)

type FederationHandler struct {
	DB      *pgxpool.Pool
	Storage *services.StorageService
}

// GET /api/fed/users?username=alice  (no auth, public)
// @Summary      Look up a local user (federation)
// @Description  Called by remote Kutup servers during federated sharing. Rate-limited to 60/min per IP.
// @Tags         Federation
// @Produce      json
// @Param        username  query     string  true  "Username to look up"
// @Success      200       {object}  PubkeyResponse
// @Failure      400       {object}  ErrorResponse
// @Failure      404       {object}  ErrorResponse
// @Failure      429       {object}  ErrorResponse  "Rate limited"
// @Router       /fed/users [get]
func (h *FederationHandler) GetUserByUsername(c *fiber.Ctx) error {
	username := c.Query("username")
	if username == "" {
		return c.Status(400).JSON(fiber.Map{"error": "username required"})
	}
	var publicKey string
	err := h.DB.QueryRow(context.Background(),
		`SELECT public_key FROM users WHERE username = $1 AND is_active = true`,
		username,
	).Scan(&publicKey)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "user not found"})
	}
	return c.JSON(fiber.Map{"publicKey": publicKey})
}

// GET /api/fed/invites/{token}  (no auth - token is the auth)
// @Summary      Get federated invite metadata
// @Tags         Federation
// @Produce      json
// @Param        token  path      string  true  "Invite token"
// @Success      200    {object}  FedInviteResponse
// @Failure      404    {object}  ErrorResponse
// @Router       /fed/invites/{token} [get]
func (h *FederationHandler) GetInvite(c *fiber.Ctx) error {
	token := c.Params("token")
	var row struct {
		EncryptedCollectionKey string
		EncryptedName          string
		NameNonce              string
		CanUpload              bool
		CanDelete              bool
		UploadQuotaBytes       *int64
	}
	err := h.DB.QueryRow(context.Background(), `
		SELECT fos.encrypted_collection_key,
		       c.encrypted_name, c.name_nonce,
		       fos.can_upload, fos.can_delete, fos.upload_quota_bytes
		FROM federated_outgoing_shares fos
		JOIN collections c ON c.id = fos.collection_id
		WHERE fos.access_token = $1
	`, token).Scan(
		&row.EncryptedCollectionKey, &row.EncryptedName, &row.NameNonce,
		&row.CanUpload, &row.CanDelete, &row.UploadQuotaBytes,
	)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "invite not found"})
	}
	return c.JSON(fiber.Map{
		"wrappedKey":       row.EncryptedCollectionKey,
		"encryptedName":    row.EncryptedName,
		"nameNonce":        row.NameNonce,
		"canUpload":        row.CanUpload,
		"canDelete":        row.CanDelete,
		"uploadQuotaBytes": row.UploadQuotaBytes,
	})
}

// GET /api/fed/shares/{token}/files  (token is auth)
// @Summary      List files in a federated share
// @Tags         Federation
// @Produce      json
// @Param        token  path      string  true  "Share access token"
// @Success      200    {array}   FileRow
// @Failure      403    {object}  ErrorResponse
// @Router       /fed/shares/{token}/files [get]
func (h *FederationHandler) ListShareFiles(c *fiber.Ctx) error {
	token := c.Params("token")
	var collectionID string
	err := h.DB.QueryRow(context.Background(),
		`SELECT collection_id FROM federated_outgoing_shares WHERE access_token = $1`,
		token,
	).Scan(&collectionID)
	if err != nil {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}

	rows, err := h.DB.Query(context.Background(), `
		SELECT id, collection_id, uploader_user_id,
		       encrypted_metadata, metadata_nonce,
		       encrypted_file_key, file_key_nonce,
		       encrypted_size_bytes, created_at, updated_at
		FROM files WHERE collection_id = $1 ORDER BY created_at DESC
	`, collectionID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	defer rows.Close()

	type FileRow struct {
		ID                 string    `json:"id"`
		CollectionID       string    `json:"collectionId"`
		UploaderUserID     string    `json:"uploaderUserId"`
		EncryptedMetadata  string    `json:"encryptedMetadata"`
		MetadataNonce      string    `json:"metadataNonce"`
		EncryptedFileKey   string    `json:"encryptedFileKey"`
		FileKeyNonce       string    `json:"fileKeyNonce"`
		EncryptedSizeBytes int64     `json:"encryptedSizeBytes"`
		CreatedAt          time.Time `json:"createdAt"`
		UpdatedAt          time.Time `json:"updatedAt"`
	}

	var files []FileRow
	for rows.Next() {
		var f FileRow
		if err := rows.Scan(&f.ID, &f.CollectionID, &f.UploaderUserID,
			&f.EncryptedMetadata, &f.MetadataNonce,
			&f.EncryptedFileKey, &f.FileKeyNonce,
			&f.EncryptedSizeBytes, &f.CreatedAt, &f.UpdatedAt); err != nil {
			continue
		}
		files = append(files, f)
	}
	if files == nil {
		files = []FileRow{}
	}
	return c.JSON(files)
}

// GET /api/fed/shares/{token}/files/{fileId}/download
// @Summary      Download a file from a federated share
// @Tags         Federation
// @Produce      octet-stream
// @Param        token   path  string  true  "Share access token"
// @Param        fileId  path  string  true  "File UUID"
// @Success      200
// @Failure      403  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Router       /fed/shares/{token}/files/{fileId}/download [get]
func (h *FederationHandler) DownloadShareFile(c *fiber.Ctx) error {
	token := c.Params("token")
	fileID := c.Params("fileId")

	// Validate token
	var collectionID string
	err := h.DB.QueryRow(context.Background(),
		`SELECT collection_id FROM federated_outgoing_shares WHERE access_token = $1`,
		token,
	).Scan(&collectionID)
	if err != nil {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}

	// Get file
	var storagePath string
	var fileSize int64
	err = h.DB.QueryRow(context.Background(),
		`SELECT storage_path, encrypted_size_bytes FROM files WHERE id = $1 AND collection_id = $2`,
		fileID, collectionID,
	).Scan(&storagePath, &fileSize)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	body, size, err := h.Storage.GetObject(c.Context(), storagePath)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	c.Set("Content-Type", "application/octet-stream")
	c.Set("Content-Length", fmt.Sprintf("%d", size))
	return c.SendStream(body, int(size))
}

// POST /api/fed/shares/{token}/files  (upload via federation)
// @Summary      Upload a file to a federated share
// @Tags         Federation
// @Accept       mpfd
// @Produce      json
// @Param        token              path      string  true  "Share access token"
// @Param        encryptedMetadata  formData  string  true  "Encrypted metadata (base64)"
// @Param        metadataNonce      formData  string  true  "Metadata nonce (base64)"
// @Param        encryptedFileKey   formData  string  true  "Encrypted file key (base64)"
// @Param        fileKeyNonce       formData  string  true  "File key nonce (base64)"
// @Param        file               formData  file    true  "Encrypted file content"
// @Success      201  {object}  UploadResult
// @Failure      400  {object}  ErrorResponse
// @Failure      403  {object}  ErrorResponse
// @Failure      413  {object}  ErrorResponse
// @Router       /fed/shares/{token}/files [post]
func (h *FederationHandler) UploadShareFile(c *fiber.Ctx) error {
	token := c.Params("token")

	// Validate token and check can_upload
	var shareID, collectionID, sharerUserID string
	var canUpload bool
	var uploadQuotaBytes *int64
	err := h.DB.QueryRow(context.Background(), `
		SELECT id, collection_id, sharer_user_id, can_upload, upload_quota_bytes
		FROM federated_outgoing_shares WHERE access_token = $1
	`, token).Scan(&shareID, &collectionID, &sharerUserID, &canUpload, &uploadQuotaBytes)
	if err != nil {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}
	if !canUpload {
		return c.Status(403).JSON(fiber.Map{"error": "upload not permitted"})
	}

	// Parse multipart
	form, err := c.MultipartForm()
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid multipart form"})
	}
	encMetadata := firstField(form.Value, "encryptedMetadata")
	metadataNonce := firstField(form.Value, "metadataNonce")
	encFileKey := firstField(form.Value, "encryptedFileKey")
	fileKeyNonce := firstField(form.Value, "fileKeyNonce")

	fileHeaders := form.File["file"]
	if len(fileHeaders) == 0 {
		return c.Status(400).JSON(fiber.Map{"error": "no file provided"})
	}
	fileHeader := fileHeaders[0]
	fileSize := fileHeader.Size

	fileID := uuid.New().String()
	storagePath := fmt.Sprintf("fed/%s/%s/%s", shareID, collectionID, fileID)

	src, err := fileHeader.Open()
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "cannot read file"})
	}
	defer src.Close()

	if err := h.Storage.Upload(c.Context(), storagePath, src, fileSize); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "storage error"})
	}

	// S3-5: Use a transaction with FOR UPDATE to atomically check and update
	// the federated quota, preventing concurrent uploads from bypassing it.
	tx, err := h.DB.Begin(c.Context())
	if err != nil {
		h.Storage.Delete(context.Background(), storagePath)
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	defer tx.Rollback(c.Context())

	if uploadQuotaBytes != nil {
		var used int64
		tx.QueryRow(c.Context(),
			`SELECT COALESCE(upload_used_bytes, 0) FROM federated_outgoing_shares WHERE id = $1 FOR UPDATE`,
			shareID,
		).Scan(&used)
		if used+fileSize > *uploadQuotaBytes {
			h.Storage.Delete(context.Background(), storagePath)
			return c.Status(413).JSON(fiber.Map{"error": "share quota exceeded"})
		}
	}

	// Insert file record (uploader is the sharer user - best proxy; actual remote user unknown)
	_, err = tx.Exec(c.Context(), `
		INSERT INTO files (id, collection_id, uploader_user_id,
		                   encrypted_metadata, metadata_nonce,
		                   encrypted_file_key, file_key_nonce,
		                   storage_path, encrypted_size_bytes)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
	`, fileID, collectionID, sharerUserID,
		encMetadata, metadataNonce,
		encFileKey, fileKeyNonce,
		storagePath, fileSize)
	if err != nil {
		h.Storage.Delete(context.Background(), storagePath)
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	// Update upload_used_bytes on the share
	tx.Exec(c.Context(),
		`UPDATE federated_outgoing_shares SET upload_used_bytes = upload_used_bytes + $1 WHERE id = $2`,
		fileSize, shareID)

	// Update sharer's storage quota
	tx.Exec(c.Context(),
		`UPDATE users SET storage_used_bytes = storage_used_bytes + $1 WHERE id = $2`,
		fileSize, sharerUserID)

	if err := tx.Commit(c.Context()); err != nil {
		h.Storage.Delete(context.Background(), storagePath)
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	return c.Status(201).JSON(fiber.Map{"id": fileID})
}

// DELETE /api/fed/shares/{token}/files/{fileId}
// @Summary      Delete a file from a federated share
// @Tags         Federation
// @Param        token   path  string  true  "Share access token"
// @Param        fileId  path  string  true  "File UUID"
// @Success      204
// @Failure      403  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Router       /fed/shares/{token}/files/{fileId} [delete]
func (h *FederationHandler) DeleteShareFile(c *fiber.Ctx) error {
	token := c.Params("token")
	fileID := c.Params("fileId")

	var shareID, collectionID, sharerUserID string
	var canDelete bool
	err := h.DB.QueryRow(context.Background(), `
		SELECT id, collection_id, sharer_user_id, can_delete
		FROM federated_outgoing_shares WHERE access_token = $1
	`, token).Scan(&shareID, &collectionID, &sharerUserID, &canDelete)
	if err != nil {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}
	if !canDelete {
		return c.Status(403).JSON(fiber.Map{"error": "delete not permitted"})
	}

	var storagePath string
	var fileSize int64
	err = h.DB.QueryRow(context.Background(),
		`SELECT storage_path, encrypted_size_bytes FROM files WHERE id = $1 AND collection_id = $2`,
		fileID, collectionID,
	).Scan(&storagePath, &fileSize)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	_, err = h.DB.Exec(context.Background(), `DELETE FROM files WHERE id = $1`, fileID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	h.DB.Exec(context.Background(),
		`UPDATE federated_outgoing_shares SET upload_used_bytes = GREATEST(0, upload_used_bytes - $1) WHERE id = $2`,
		fileSize, shareID)
	h.DB.Exec(context.Background(),
		`UPDATE users SET storage_used_bytes = GREATEST(0, storage_used_bytes - $1) WHERE id = $2`,
		fileSize, sharerUserID)
	h.Storage.Delete(context.Background(), storagePath)

	return c.SendStatus(204)
}

File: crates/kutup-server/src/handlers/federation.rs
//! Federation public endpoints — mirrors `backend/handlers/federation.go`.
//!
//! These are the endpoints a *remote* kutup server calls during federated sharing. There is
//! no JWT here — the per-share `access_token` is the capability. The server still only ever
//! handles ciphertext (encrypted metadata + file blobs).

use std::io::Write;

use aws_sdk_s3::primitives::ByteStream;
use axum::extract::{Multipart, Path, Query, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::{Deserialize, Serialize};
use serde_json::json;
use tempfile::NamedTempFile;
use time::OffsetDateTime;
use uuid::Uuid;

use crate::error::{AppError, AppResult};
use crate::handlers::octet_stream_response;
use crate::AppState;

#[derive(Debug, Deserialize)]
pub struct UserQuery {
    username: Option<String>,
}

/// `GET /api/fed/users?username=…` — mirrors `GetUserByUsername`. Rate-limited (60/min/IP)
/// by the route layer. Returns the recipient's public key for wrapping the collection key.
pub async fn get_user_by_username(
    State(state): State<AppState>,
    Query(q): Query<UserQuery>,
) -> AppResult<Response> {
    let username = q.username.unwrap_or_default();
    if username.is_empty() {
        return Err(AppError::bad_request("username required"));
    }
    let pubkey: Option<String> =
        sqlx::query_scalar("SELECT public_key FROM users WHERE username = $1 AND is_active = true")
            .bind(&username)
            .fetch_optional(&state.pool)
            .await
            .ok()
            .flatten();
    match pubkey {
        Some(pk) => Ok(Json(json!({ "publicKey": pk })).into_response()),
        None => Err(AppError::not_found("user not found")),
    }
}

/// Invite metadata for a federated share. Keys are alphabetical, matching Go's marshalled
/// `fiber.Map`.
#[derive(Debug, Serialize)]
struct InviteResponse {
    #[serde(rename = "canDelete")]
    can_delete: bool,
    #[serde(rename = "canUpload")]
    can_upload: bool,
    #[serde(rename = "encryptedName")]
    encrypted_name: String,
    #[serde(rename = "nameNonce")]
    name_nonce: String,
    #[serde(rename = "uploadQuotaBytes")]
    upload_quota_bytes: Option<i64>,
    #[serde(rename = "wrappedKey")]
    wrapped_key: String,
}

/// `GET /api/fed/invites/{token}` — mirrors `GetInvite`. The token is the auth.
pub async fn get_invite(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> AppResult<Response> {
    let row: Option<(String, String, String, bool, bool, Option<i64>)> = sqlx::query_as(
        r#"SELECT fos.encrypted_collection_key, c.encrypted_name, c.name_nonce,
                  fos.can_upload, fos.can_delete, fos.upload_quota_bytes
           FROM federated_outgoing_shares fos
           JOIN collections c ON c.id = fos.collection_id
           WHERE fos.access_token = $1"#,
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((wrapped_key, encrypted_name, name_nonce, can_upload, can_delete, quota)) = row else {
        return Err(AppError::not_found("invite not found"));
    };
    Ok(Json(InviteResponse {
        can_delete,
        can_upload,
        encrypted_name,
        name_nonce,
        upload_quota_bytes: quota,
        wrapped_key,
    })
    .into_response())
}

/// One file in a federated share. Field order mirrors the Go struct.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct FedFileRow {
    id: Uuid,
    collection_id: Uuid,
    uploader_user_id: Uuid,
    encrypted_metadata: String,
    metadata_nonce: String,
    encrypted_file_key: String,
    file_key_nonce: String,
    encrypted_size_bytes: i64,
    #[serde(with = "time::serde::rfc3339")]
    created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    updated_at: OffsetDateTime,
}

/// `GET /api/fed/shares/{token}/files` — mirrors `ListShareFiles`.
pub async fn list_share_files(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> AppResult<Response> {
    let coll_id: Option<Uuid> = sqlx::query_scalar(
        "SELECT collection_id FROM federated_outgoing_shares WHERE access_token = $1",
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some(coll_id) = coll_id else {
        return Err(AppError::forbidden("forbidden"));
    };

    type FedFileTuple = (
        Uuid,
        Uuid,
        Uuid,
        String,
        String,
        String,
        String,
        i64,
        OffsetDateTime,
        OffsetDateTime,
    );
    let rows: Vec<FedFileTuple> = sqlx::query_as(
        r#"SELECT id, collection_id, uploader_user_id, encrypted_metadata, metadata_nonce,
                  encrypted_file_key, file_key_nonce, encrypted_size_bytes, created_at, updated_at
           FROM files WHERE collection_id = $1 ORDER BY created_at DESC"#,
    )
    .bind(coll_id)
    .fetch_all(&state.pool)
    .await
    .map_err(|_| AppError::internal("internal error"))?;

    let files: Vec<FedFileRow> = rows
        .into_iter()
        .map(
            |(id, cid, uid, em, mn, efk, fkn, size, created, updated)| FedFileRow {
                id,
                collection_id: cid,
                uploader_user_id: uid,
                encrypted_metadata: em,
                metadata_nonce: mn,
                encrypted_file_key: efk,
                file_key_nonce: fkn,
                encrypted_size_bytes: size,
                created_at: created,
                updated_at: updated,
            },
        )
        .collect();
    Ok(Json(files).into_response())
}

/// `GET /api/fed/shares/{token}/files/{fileId}/download` — mirrors `DownloadShareFile`.
pub async fn download_share_file(
    State(state): State<AppState>,
    Path((token, file_id)): Path<(String, String)>,
) -> AppResult<Response> {
    let coll_id: Option<Uuid> = sqlx::query_scalar(
        "SELECT collection_id FROM federated_outgoing_shares WHERE access_token = $1",
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some(coll_id) = coll_id else {
        return Err(AppError::forbidden("forbidden"));
    };
    let fid = Uuid::parse_str(&file_id).map_err(|_| AppError::not_found("not found"))?;
    let row: Option<(String, i64)> = sqlx::query_as(
        "SELECT storage_path, encrypted_size_bytes FROM files WHERE id = $1 AND collection_id = $2",
    )
    .bind(fid)
    .bind(coll_id)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((storage_path, _size)) = row else {
        return Err(AppError::not_found("not found"));
    };
    let (body, size) = state
        .storage
        .get_object(&storage_path)
        .await
        .map_err(|_| AppError::internal("internal error"))?;
    Ok(octet_stream_response(body, size, &[]))
}

/// `POST /api/fed/shares/{token}/files` — mirrors `UploadShareFile`. Multipart upload from a
/// remote server; the file blob is stored under `fed/{shareId}/{collectionId}/{fileId}`.
pub async fn upload_share_file(
    State(state): State<AppState>,
    Path(token): Path<String>,
    mut multipart: Multipart,
) -> AppResult<Response> {
    let share: Option<(Uuid, Uuid, Uuid, bool, Option<i64>)> = sqlx::query_as(
        r#"SELECT id, collection_id, sharer_user_id, can_upload, upload_quota_bytes
           FROM federated_outgoing_shares WHERE access_token = $1"#,
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((share_id, coll_id, sharer_id, can_upload, share_quota)) = share else {
        return Err(AppError::forbidden("forbidden"));
    };
    if !can_upload {
        return Err(AppError::forbidden("upload not permitted"));
    }

    // Collect the form fields + spool the file part to a temp file (to learn its size).
    let mut enc_metadata = String::new();
    let mut metadata_nonce = String::new();
    let mut enc_file_key = String::new();
    let mut file_key_nonce = String::new();
    let mut tmp: Option<(NamedTempFile, i64)> = None;
    loop {
        let field = multipart
            .next_field()
            .await
            .map_err(|_| AppError::bad_request("invalid multipart form"))?;
        let Some(mut field) = field else { break };
        match field.name() {
            Some("encryptedMetadata") => enc_metadata = field.text().await.unwrap_or_default(),
            Some("metadataNonce") => metadata_nonce = field.text().await.unwrap_or_default(),
            Some("encryptedFileKey") => enc_file_key = field.text().await.unwrap_or_default(),
            Some("fileKeyNonce") => file_key_nonce = field.text().await.unwrap_or_default(),
            Some("file") => {
                let mut f = NamedTempFile::new().map_err(|_| AppError::internal("temp file"))?;
                let mut size: i64 = 0;
                while let Some(chunk) = field
                    .chunk()
                    .await
                    .map_err(|_| AppError::bad_request("no file provided"))?
                {
                    f.write_all(&chunk)
                        .map_err(|_| AppError::internal("temp write"))?;
                    size += chunk.len() as i64;
                }
                tmp = Some((f, size));
            }
            _ => {}
        }
    }
    let Some((tmp_file, file_size)) = tmp else {
        return Err(AppError::bad_request("no file provided"));
    };

    let file_id = Uuid::new_v4();
    let storage_path = format!("fed/{share_id}/{coll_id}/{file_id}");

    // Upload to S3 first; on any subsequent failure we delete the orphan.
    let body = ByteStream::from_path(tmp_file.path())
        .await
        .map_err(|_| AppError::internal("read upload"))?;
    if state
        .storage
        .upload(&storage_path, body, file_size)
        .await
        .is_err()
    {
        return Err(AppError::internal("storage error"));
    }

    // FOR UPDATE on the share row to make the quota check + counter bump atomic.
    let mut tx = match state.pool.begin().await {
        Ok(t) => t,
        Err(_) => {
            let _ = state.storage.delete(&storage_path).await;
            return Err(AppError::internal("internal error"));
        }
    };
    if let Some(quota) = share_quota {
        let used: i64 = sqlx::query_scalar(
            "SELECT COALESCE(upload_used_bytes, 0) FROM federated_outgoing_shares WHERE id = $1 FOR UPDATE",
        )
        .bind(share_id)
        .fetch_one(&mut *tx)
        .await
        .unwrap_or(0);
        if used + file_size > quota {
            let _ = state.storage.delete(&storage_path).await;
            return Err(AppError::new(
                StatusCode::PAYLOAD_TOO_LARGE,
                "share quota exceeded",
            ));
        }
    }

    let ins = sqlx::query(
        r#"INSERT INTO files (id, collection_id, uploader_user_id,
                              encrypted_metadata, metadata_nonce,
                              encrypted_file_key, file_key_nonce,
                              storage_path, encrypted_size_bytes)
           VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)"#,
    )
    .bind(file_id)
    .bind(coll_id)
    .bind(sharer_id)
    .bind(&enc_metadata)
    .bind(&metadata_nonce)
    .bind(&enc_file_key)
    .bind(&file_key_nonce)
    .bind(&storage_path)
    .bind(file_size)
    .execute(&mut *tx)
    .await;
    if ins.is_err() {
        let _ = state.storage.delete(&storage_path).await;
        return Err(AppError::internal("internal error"));
    }

    let _ = sqlx::query(
        "UPDATE federated_outgoing_shares SET upload_used_bytes = upload_used_bytes + $1 WHERE id = $2",
    )
    .bind(file_size)
    .bind(share_id)
    .execute(&mut *tx)
    .await;
    let _ =
        sqlx::query("UPDATE users SET storage_used_bytes = storage_used_bytes + $1 WHERE id = $2")
            .bind(file_size)
            .bind(sharer_id)
            .execute(&mut *tx)
            .await;

    if tx.commit().await.is_err() {
        let _ = state.storage.delete(&storage_path).await;
        return Err(AppError::internal("internal error"));
    }

    Ok((StatusCode::CREATED, Json(json!({ "id": file_id }))).into_response())
}

/// `DELETE /api/fed/shares/{token}/files/{fileId}` — mirrors `DeleteShareFile`.
pub async fn delete_share_file(
    State(state): State<AppState>,
    Path((token, file_id)): Path<(String, String)>,
) -> AppResult<Response> {
    let share: Option<(Uuid, Uuid, Uuid, bool)> = sqlx::query_as(
        r#"SELECT id, collection_id, sharer_user_id, can_delete
           FROM federated_outgoing_shares WHERE access_token = $1"#,
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((share_id, coll_id, sharer_id, can_delete)) = share else {
        return Err(AppError::forbidden("forbidden"));
    };
    if !can_delete {
        return Err(AppError::forbidden("delete not permitted"));
    }
    let fid = Uuid::parse_str(&file_id).map_err(|_| AppError::not_found("not found"))?;
    let row: Option<(String, i64)> = sqlx::query_as(
        "SELECT storage_path, encrypted_size_bytes FROM files WHERE id = $1 AND collection_id = $2",
    )
    .bind(fid)
    .bind(coll_id)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((storage_path, file_size)) = row else {
        return Err(AppError::not_found("not found"));
    };

    if sqlx::query("DELETE FROM files WHERE id = $1")
        .bind(fid)
        .execute(&state.pool)
        .await
        .is_err()
    {
        return Err(AppError::internal("internal error"));
    }
    let _ = sqlx::query(
        "UPDATE federated_outgoing_shares SET upload_used_bytes = GREATEST(0, upload_used_bytes - $1) WHERE id = $2",
    )
    .bind(file_size)
    .bind(share_id)
    .execute(&state.pool)
    .await;
    let _ = sqlx::query(
        "UPDATE users SET storage_used_bytes = GREATEST(0, storage_used_bytes - $1) WHERE id = $2",
    )
    .bind(file_size)
    .bind(sharer_id)
    .execute(&state.pool)
    .await;
    let _ = state.storage.delete(&storage_path).await;

    Ok(StatusCode::NO_CONTENT.into_response())
}
